src/lib/frontend/frontend.ml

```diff
@@ -246,27 +270,33 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
       (* This case should mainly occur when a query has a non-unsat result,
          so we want to print the status in this case. *)
       print_status (Sat (d,t)) (Steps.get_steps ());
-      (*if get_model () then SAT.print_model ~header:true (get_fmt_mdl ()) t;*)
-      env, `Sat t, dep
+      print_model env (Some SAT.NoTimeout);
+      env , consistent, dep
     | SAT.Unsat dep' ->
       (* This case should mainly occur when a new assumption results in an unsat
          env, in which case we do not want to print status, since the correct
          status should be printed at the next query. *)
       let dep = Ex.union dep dep' in
       if get_debug_unsat_core () then check_produced_unsat_core dep;
       (* print_status (Inconsistent d) (Steps.get_steps ()); *)
-      env, `Unsat, dep
-    | SAT.I_dont_know t ->
-      (* In this case, it's not clear whether we want to print the status.
-         Instead, it'd be better to accumulate in `consistent` a 3-case adt
-         and not a simple bool. *)
-      print_status (Unknown (d, t)) (Steps.get_steps ());
-      (*if get_model () then SAT.print_model ~header:true (get_fmt_mdl ()) t;*)
-      env, `Unknown t, dep
+      env , `Unsat, dep
+    | SAT.I_dont_know {env; timeout} ->
```

Comment:
I agree this shadowing is misleading but we have to return in this branch the SAT environment fron the caught exception:
```ocaml
    | SAT.I_dont_know {env = t; timeout} ->
      (* TODO: always print Unknown for why3 ? *)
      let status =
        if timeout != NoTimeout then (Timeout (Some d))
        else (Unknown (d, t))
      in
      print_status status (Steps.get_steps ());
      print_model t (Some timeout);
      (* TODO: Is it an appropriate behaviour? *)
      (*       if timeout != NoTimeout then raise Util.Timeout; *)
      env, `Unknown t, dep

```

Reply:
Yeah, the first element of the returned triple `env, `Unknown t, dep` is `env` which is shadowed I believe? From what I remember the first element needs to be `env` here for correctness (otherwise the env has additional assertions that are not part of the semantic) but I may be wrong

Reply:
Oh you are right! The previous code was wrong beceause of the shadowing. Now it's fixed. We should add some tests for push and pop, otherwise we cannot detect this kind of mistake easily.